### ISSUE TITLE
files.uploadFile() not working with custom stream

### ISSUE THREAD

**Issue by F145h** [box/box-node-sdk#463]
Have a generated text content on heroku(without write access to the filesystem), upload to box.com failed with error "Unexpected API Response [400 Bad Request] bad_request - Stream ended unexpectedly"

sample:
```
                const readable = new Readable();
                readable._read = () => {}

                readable.push(Buffer.from("Hello", 'utf8'));
                readable.push(null);

                boxConnection.files.uploadFile("0", "test.txt", readable).then(file => {
                     ....
                }
```

1) getChunkedUploader() works
2) empty stream( without `readable.push(Buffer.from("Hello", 'utf8'));`) works too
3) readable.readableLength has the correct value 5 bytes

**Comment by sgalizia** [comment]
This is because Box uses `request` under the hood and the library requires that you provide a `content-length` header. On the Box community site they suggest editing the prototype of the stream to add these properties that `request` is looking for. I **STRONGLY** disagree with this. It is an underlying fault in their SDK and they need to provide a way for users to tell the SDK how big the file is that they want to upload.

I'm hoping to get their attention on this to fix it as I ran into the EXACT same issue yesterday and wasted hours trying to figure this out. When you are not using a `fs` stream it does not work. Give us the ability to use non `fs` streams and specify the size please!

**Comment by F145h** [comment]
A good solution: make POST request for `https://upload.box.com/api/2.0/files/content` and upload content as multupart. Then you need to work and updating accessToken yourself. And SDK is not needed 🤨

**Comment by PJSimon** [comment]
Hi @F145h and @sgalizia,

Thanks for bringing up this excellent feedback.  I'll discuss with the team and get back to you soon!

Patrick

**Comment by sujaygarlanka** [comment]
Hello @F145h and @sgalizia,

What is the reason you guys have for not using fs streams? We would like to know so we can appropriately prioritize this request.

**Comment by F145h** [comment]
> Hello @F145h and @sgalizia,
> 
> What is the reason you both have for not using fs streams? We would like to know so we can appropriately prioritize this request.

In my case heroku hosting does not allow write to the local file system, therefore I can’t use fs-only streams.
Well, custom stream does not work in the example as I showed, it is simple. But it works fine with dropbox, aws-sdk etc. 
Why do we need save a temporary file if the data is already have in the buffer / stream

**Comment by sgalizia** [comment]
> 
> 
> Hello @F145h and @sgalizia,
> 
> What is the reason you both have for not using fs streams? We would like to know so we can appropriately prioritize this request.

My reason is very similar to @F145h, we are streaming the file from an sftp server and writing to the local filesystem is a waste when the stream is already there as a buffer or open stream. It also slows down the process of what we are trying to do.

**Comment by stephencweiss** [comment]
I'm also running into this issue where my stream is coming from AWS (using their SDK to `getObject` from the s3 service).

I'm currently trying to avoid writing a temporary file to disk just to read it back in using `fs.createReadStream`

Happy to use chunked, but this hard limitation on needing to know the bit size up front instead of awaiting an `end` event from the stream is creating a lot of pain.

**Comment by sujaygarlanka** [comment]
This issue has been fixed by the PR #519. Will be releasing a new version with this fix soon.
